### PR TITLE
AMDGPU: Treat uint32_max as the default value for amdgpu-max-num-workgroups

### PR DESCRIPTION
**Pull request by arsenm**
0 does not make sense as a value for this to be, much less the default.
Also stop emitting each individual field if it is the default, rather than
if any element was the default. Also fix the name of the test since it didn't
exactly match the real attribute name.

### CODE REVIEW

**Comment by arsenm** [comment]
* **#113038** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/113038?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#113019** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/113019?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#113018** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/113018?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#113751** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/113751?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @arsenm and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: llvm/test/CodeGen/AMDGPU/attr-amdgpu-max-num-workgroups.ll [hunk 1]
```diff
@@ -54,16 +81,25 @@ attributes #4 = {"amdgpu-max-num-workgroups"="1024,1024,1024"}
 
 ; CHECK: - .args:
 ; CHECK:        .max_flat_workgroup_size: 1024
+; CHECK-NEXT:   .max_num_workgroups_x: 0
```
Comment (shiltian) [inline llvm/test/CodeGen/AMDGPU/attr-amdgpu-max-num-workgroups.ll:84]: What does 0 here mean?

Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/attr-amdgpu-max-num-workgroups.ll:84]: It doesn't say that. It states "Clang only emits this attribute when all the three numbers are >= 1." Which doesn't really make sense.

0 doesn't make sense as a value

Comment (shiltian) [reply llvm/test/CodeGen/AMDGPU/attr-amdgpu-max-num-workgroups.ll:84]: It does say that: "The max number of launched work-groups in the X, Y, and Z dimensions. Each number must be >=1."

Comment (krzysz00) [reply llvm/test/CodeGen/AMDGPU/attr-amdgpu-max-num-workgroups.ll:84]: In trivial questions, is the hardware maximum for workgroups actually `uint32_t_max` in all dimensions, or is it `uint32_t_max / [wave size]`?

Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/attr-amdgpu-max-num-workgroups.ll:84]: That's on the emitted directive, it's not directly documented on the IR attribute.

I think I've seen that it's lower in the y and z dimensions but I don't remember where and can't find documentation for it

**Review by shiltian — APPROVED** [review]

File: llvm/docs/AMDGPUUsage.rst [hunk 2]
```diff
@@ -1645,9 +1645,9 @@ The AMDGPU backend supports the following LLVM IR attributes.
                                              reduced by heuristics.
 
      "amdgpu-max-num-workgroups"="x,y,z"     Specify the maximum number of work groups for the kernel dispatch in the
-                                             X, Y, and Z dimensions. Generated by the ``amdgpu_max_num_work_groups``
-                                             CLANG attribute [CLANG-ATTR]_. Clang only emits this attribute when all
-                                             the three numbers are >= 1.
+                                             X, Y, and Z dimensions. Each number must be >= 1. Generated by the
+                                             ``amdgpu_max_num_work_groups`` CLANG attribute [CLANG-ATTR]_. Clang only
+                                             emits this attribute when all the three numbers are >= 1.
```
Comment (shiltian) [inline llvm/docs/AMDGPUUsage.rst:1650]: > clang only emits this attribute when all the three numbers are >= 1.

Do we have sema check for this?

Comment (arsenm) [reply llvm/docs/AMDGPUUsage.rst:1650]: Should, that part isn't new

**Pull request merged**